recipes/integrate-social-logins.md

```diff
@@ -0,0 +1,581 @@
+6. The API finally create its own token and send it back to the original popup by redirecting to **www**.example.com/receive?token=abcdef
+7. The popup find the token in the URL and forward it to the popup opener via a JavaScript function
+8. Finally the popup close itself, leaving only the original login page which is now in possession of a token that can be used for any subsequent API calls
+9. Optionally, the login page could store the token in local storage, so the login process can be avoided the next time the user visit the site
+
+Implementation of login
+-----------------------
+
+### Server side
+
+First of all install HybridAuth which we'll use to talk to third party providers:
+
+``composer require  hybridauth/hybridauth:dev-3.0.0-Remake``
+
+Then create a new RPC service ``HybridAuth`` via Apigility admin interface. Since we will need two action in the same controller, we'll define the route as ``/hybridauth[/:action]``:
+
+![Create a new RPC service](/asset/apigility-documentation/img/recipes-social-login-rpc-service.png)
+
+In ``HybridAuthController.php``, we create two actions. The first one, ``hybridAuthAction()`` will redirect to the provider. And the second action, ``callbackAction()`` will be used as a callback for the provider. This is where we will create/update user, generate a token, and give it the popup.
+
+```php
+<?php
+
+class HybridAuthController extends AbstractActionController
+{
+    use \Application\Traits\EntityManagerAware;
+
+    /**
+     * Returns the select authentification provider
+     * @return \Hybridauth\Adapter\AdapterInterface
+     */
+    private function getProvider()
+    {
+        $uri = $this->getRequest()->getUri();
+        $base = sprintf('%s://%s', $uri->getScheme(), $uri->getHost());
+
+        $provider = $this->getRequest()->getQuery('provider');
+        $config = $this->getServiceLocator()->get('Config')['hybridauth'];
+        $config['callback'] = $base . $this->url()->fromRoute('MY-API-NAME.rpc.hybrid-auth', ['action' => 'callback']) . '?provider=' . $provider;
+
+        $hybridauth = new \Hybridauth\Hybridauth($config);
+        $adapter = $hybridauth->getAdapter($provider);
+
+        return $adapter;
+    }
+
+    public function hybridAuthAction()
+    {
+        $provider = $this->getProvider();
+        $provider->disconnect();
+        $provider->authenticate();
+
+        // If we reach this point, that means we were already authenticated by
+        // the provider. So we finish as a normal subscribe/login
+        return $this->callbackAction();
+    }
+
+    public function callbackAction()
+    {
+        $provider = $this->getProvider();
+        $provider->authenticate();
+        $profile = $provider->getUserProfile();
+        $providerName = strtolower($this->getRequest()->getQuery('provider'));
+
+        $user = $this->getServiceLocator()->get('Doctrine\ORM\EntityManager')->getRepository(\Application\Model\User::class)->createOrUpdate($providerName, $profile);
```

Comment:
As said above, prefer inject dependencies rather than invoking ServiceManager within the actions of a Controller.
Another mark for service layer